### PR TITLE
Kernel+Tests: Handful of fixes and tests for the rmdir() syscall

### PR DESCRIPTION
**Pull request by awesomekling**
While cleaning out some ancient FIXMEs, I ran into some other issues with `rmdir()` so here are a bunch of fixes for that. :^)

### CODE REVIEW

File: Tests/LibC/TestIo.cpp [hunk 1]
```diff
@@ -317,6 +318,88 @@ TEST_CASE(tmpfs_massive_file)
...
 }
 
+TEST_CASE(rmdir_dot)
+{
+    int rc = mkdir("/home/anon/rmdir-test-1", 0700);
+    VERIFY(rc == 0);
+
+    rc = rmdir("/home/anon/rmdir-test-1/.");
+    VERIFY(rc != 0);
+    VERIFY(errno == EINVAL);
```
Comment (FalseHonesty) [inline Tests/LibC/TestIo.cpp:328]: Should these be `EXPECT`/`EXPECT_EQ`?

File: Tests/LibC/TestIo.cpp [hunk 2]
```diff
@@ -317,6 +318,88 @@ TEST_CASE(tmpfs_massive_file)
...
+    VERIFY(errno == ENOTEMPTY);
+
+    rc = rmdir("/home/anon/rmdir-test-2/foo");
+    VERIFY(rc == 0);
+
+    rc = rmdir("/home/anon/rmdir-test-2");
+    VERIFY(rc == 0);
+}
+
+TEST_CASE(rmdir_someone_elses_file_in_my_sticky_directory)
```
Comment (tmapes) [inline Tests/LibC/TestIo.cpp:360]: Nitpick  - `rmdir_someone_elses_directory_in_my_sticky_directory` :^)

File: Kernel/FileSystem/VirtualFileSystem.cpp [hunk 3]
```diff
@@ -855,8 +855,13 @@ ErrorOr<void> VirtualFileSystem::rmdir(Credentials const& credentials, StringVie
         return EACCES;
 
     if (parent_metadata.is_sticky()) {
-        if (!credentials.is_superuser() && inode.metadata().uid != credentials.euid())
+        // [EACCES] The S_ISVTX flag is set on the directory containing the file referred to by the path argument
+        //          and the process does not satisfy the criteria specified in XBD Directory Protection.
+        if (!credentials.is_superuser()
+            && inode.metadata().uid != credentials.euid()
+            && parent_inode.metadata().uid != credentials.euid()) {
             return EACCES;
```
Comment (krkk) [inline Kernel/FileSystem/VirtualFileSystem.cpp:863]: probably out of scope, but the logic should also be updated for `rename()` and `unlink()`

File: Kernel/FileSystem/VirtualFileSystem.cpp [hunk 4]
```diff
@@ -822,30 +822,46 @@ ErrorOr<void> VirtualFileSystem::symlink(Credentials const& credentials, StringV
...
     if (!parent_metadata.may_write(credentials))
         return EACCES;
 
     if (parent_metadata.is_sticky()) {
-        if (!credentials.is_superuser() && inode.metadata().uid != credentials.euid())
+        // [EACCES] The S_ISVTX flag is set on the directory containing the file referred to by the path argument
+        //          and the process does not satisfy the criteria specified in XBD Directory Protection.
+        if (!credentials.is_superuser()
+            && inode.metadata().uid != credentials.euid()
+            && parent_inode.metadata().uid != credentials.euid()) {
```
Comment (xZise) [inline Kernel/FileSystem/VirtualFileSystem.cpp:862]:
```suggestion
            && parent_metadata().uid != credentials.euid()) {
```

There is already such a variable so why not use it?

File: Kernel/FileSystem/VirtualFileSystem.cpp (hunk 3 above)
Comment (awesomekling) [reply Kernel/FileSystem/VirtualFileSystem.cpp:863]: Yeah let's do those separately. Wanna make a PR? :^)

Comment (krkk) [reply Kernel/FileSystem/VirtualFileSystem.cpp:863]: yeah why not ¯\\_(ツ)_/¯. will probably try to revive the whole #15093 then

File: Kernel/FileSystem/VirtualFileSystem.cpp [hunk 5]
```diff
@@ -854,9 +870,12 @@ ErrorOr<void> VirtualFileSystem::rmdir(Credentials const& credentials, StringVie
         return {};
```
Comment (jdahlin) [inline Kernel/FileSystem/VirtualFileSystem.cpp:870]: Slightly unrelated, but I it should be possible to terminate the count once it reached 3.

File: Kernel/FileSystem/VirtualFileSystem.cpp [hunk 6]
```diff
@@ -822,30 +822,46 @@ ErrorOr<void> VirtualFileSystem::symlink(Credentials const& credentials, StringV
...
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/rmdir.html
 ErrorOr<void> VirtualFileSystem::rmdir(Credentials const& credentials, StringView path, Custody& base)
 {
     RefPtr<Custody> parent_custody;
     auto custody = TRY(resolve_path(credentials, path, base, &parent_custody));
     auto& inode = custody->inode();
 
-    // FIXME: We should return EINVAL if the last component of the path is "."
-    // FIXME: We should return ENOTEMPTY if the last component of the path is ".."
+    auto last_component = KLexicalPath::basename(path);
```
Comment (jdahlin) [inline Kernel/FileSystem/VirtualFileSystem.cpp:832]: Nitpick: The final return statement of the function can now be changed to

`return parent_inode.remove_child(last_componet);`

The compiler can hopefully optimize out multiple calls to `KLexicalPath::basename(path);` by itself, but I often find it easier to read code without duplicated logic.

File: Kernel/FileSystem/VirtualFileSystem.cpp (hunk 5 above)
Comment (jdahlin) [inline Kernel/FileSystem/VirtualFileSystem.cpp:870]: Unrelated, but this can probably be optimized to break out the loop when child_child > 2

**Review by supercomputer7 — APPROVED** [review]
Looks good to me :)

**Pull request merged**